### PR TITLE
FEATURE: filter API for plugins

### PR DESCRIPTION
**Pull request by SamSaffron**
This new API can be thought of as a "plugin aware" version of DiscourseEvent

When plugins register a filter it will be called automatically if plugin
is enabled and core will "filter" a parameter is sent to it using the block
the plugin registered:

```
DiscoursePluginRegistry.register_filter(plugin_instance, :magic_sum_filter) { |a, b| a + b }
sum = DiscoursePluginRegistry.apply_filter(:magic_sum_filter, 1, 2)
expect(sum).to eq(3)
```

These filters stack (first in first called)
These filters auto disable when plugin is disabled
These filters will pass the stacked result of all the block invocations to caller

### CODE REVIEW

**Review by martin-brennan — APPROVED** [review]
Looks really nice, though I will quibble with this:

> This new API can be thought of as a "plugin aware" version of DiscourseEvent

It doesn't really have anything to do conceptually with `DiscourseEvent` in my mind, since that is just a pub/sub system essentially, it's just another tool in the `DiscoursePluginRegistry` belt. Something more like:

> This new API can be used in situations where a plugin may return _different_ data than a core method, not _more_ data as is the general case for filtered_registers in `DiscoursePluginRegistry`, without the overhead of defining class-based extension points.

Maybe also mention this can be used in plugins as an alternative to overriding serializer methods?

File: lib/discourse_plugin_registry.rb [hunk 1]
```diff
@@ -246,8 +246,39 @@ def self.core_asset_for_name(name)
...
+
+  def self.apply_filter(name, arg, *more_args)
+    return arg if !@filters
+
+    filters = @filters[name]
+    return arg if !filters
+
+    # iterate as fast as possible to minimize cost (avoiding each)
+    # also erases one stack frame
+    length = filters.length
```
Comment (martin-brennan) [inline lib/discourse_plugin_registry.rb:267]: This is interesting too, I never think to do optimizations like this

File: lib/discourse_plugin_registry.rb [hunk 2]
```diff
@@ -246,8 +246,39 @@ def self.core_asset_for_name(name)
...
+  def self.register_filter(plugin_instance, name, &blk)
+    @filters ||= {}
+    filters = @filters[name] ||= []
+    filters << [plugin_instance, blk]
+  end
+
+  def self.apply_filter(name, arg, *more_args)
+    return arg if !@filters
+
+    filters = @filters[name]
```
Comment (martin-brennan) [inline lib/discourse_plugin_registry.rb:262]: A tiny nitpick, `filters` clashes too much with the instance var in my mind, `registered_filters` might be better,  but really not a huge deal

**Comment by SamSaffron** [comment]
The trouble with DiscourseEvent is that it is pretty bare bones. 

In some ways this is exactly the same system (you stack filters) 

But DiscourseEvent 

- Does not allow you to "filter/expand" incoming args - it just calls the callbacks 
- DiscourseEvent is not "plugin enabled" aware, so anyone consuming from a plugin also needs to check site settings inside the event

Changing DiscourseEvent though to suit this purpose would be somewhat risky... I think the fundamental issue though is more naming vs function... Event is all about broadcast, nothing about core consuming the value plugins produce.

**Comment by martin-brennan** [comment]
@SamSaffron I think I was unclear, I am not suggesting to do anything with `DiscourseEvent`, that is only useful for very certain cases where you are reacting to a core event in a plugin, so it will only happen if the plugin is enabled anyway. It is not useful at all for the attributes in serializers issue or many other things solved by `DiscoursePluginRegistry`. Its its own thing and fine as is.

What you've done here is right, I was just saying that calling this a "plugin-aware version of `DiscourseEvent`" does not sound correct to me, which is why I suggested that alternate description for the commit.

**Comment by SamSaffron** [comment]
yeah understood @martin-brennan ... was just trying to explain where the systems are similar and the gaps discourse event has, happy to adjust the commit message, for sure.

**Pull request merged**